content/blog/2020-12-02-chlorophyll/index.md

```diff
@@ -0,0 +1,72 @@
+
+The other challenge is that when breaking down the large program into smaller pieces, the selection of the boundary can be tricky. The authors applied superoptimization on a sliding window of superoptimization units and consecutively find valid superoptimized segments. Although dynamic programming is able to produce a better result, the sliding window technique turns out to be more efficient. 
+
+The authors used Z3 SMT solver to perform the search. The state of a program and communication channels are encoded as a large bitvector that is fed in Z3. The address space is compressed to speed up the searching process.
+
+## Limitations
+There are some occasions when the proposed solution misses some optimization opportunities or fails to produce partitions that are small enough to fit in cores. Since both partitioning and routing stages are schedule-oblivious, the output can introduce performance loss. Moreover, some context-related optimizations can be missing as the scope of superoptimization is still local to individual segments. To deal with the oversized partitions generated from the partitioning synthesizer, the compiler employs an iterative refinement method which increases the estimate margin in each rerun after the output fails to fit in cores.
+
+## Evaluation
+This paper organizes the experiments in a clear and concise manner. The benefits brought by the proposed framework are presented as several hypotheses, and experiments are designed to prove their validity. 
+
+<p align="center">
+  <img width="600" src="https://github.com/Sibylau/cs6120/blob/chlorophyll-blog/content/blog/2020-12-02-chlorophyll/1.PNG">
+</p>
+
+The above figure shows the performance improvement with sliding window superoptimization, partition synthesizer, and layout synthesizer. Another benefit observed is that the partitioning synthesizer produces smalled programs that occupy fewer cores. The authors also evaluate results against expert-written programs on several single-core applications and one multi-core MD5 hash application. The programs generated by Chlorophyll are comparable with hand-written programs. 
+
+One interesting observation is that the superoptimizer can discover clever optimizations including logic simplification, strength reduction, and CSE. With the help of Chlorophyll, programmers can easily explore different implementations and get out a design with good performance in a short period of time. And Chlorophyll can be improved by providing more human insights.
+
+## Questions
+
+1. What are the unique advantages of synthesis-aided compilers over traditional compilers?
+2. Why did the authors tear apart the compilation problem into four subproblems?
+3. Why did the authors use superoptimization instead of dynamic programming when doing code generation? What benefits can be foreseen when choosing the algorithms?

```

Comment:
Not a bad question, but I'm actually not sure how you'd use DP to do optimization for code generation. Are you thinking of a DP-based instruction selection technique?

Reply:
Yes, by mentioning DP I'm thinking about a bottom-up approach - breaking down to small instruction sequences and performing local optimizations on the way. I was actually wondering how the authors decide to go with the choice of using superoptimization in their code gen. Because superoptimization has a very obvious drawback which is its scalability. It seems to me that with superoptimization, it is still finding the local optimal.

Reply:
Great; thanks for clarifying!